File: Python/5-BinarySearch/Problems/3-FindMinimum-in-RotatedSortedArray/FindMinimum-in-RotatedSortedArray.py
# Solution for Finding Minimum in a Rotated Sorted Array
# Implementing five different approaches for various use cases

from typing import List

def find_min_iterative(nums: List[int]) -> int:
    """
    Iterative binary search approach.

    This function finds the minimum element in a rotated sorted array using an iterative
    binary search algorithm.  It's efficient, with a time complexity of O(log n).

    Args:
        nums (List[int]): A list of integers representing the rotated sorted array.

    Returns:
        int: The minimum element in the array.

    Algorithm:
        1. Initialize two pointers, `left` and `right`, to the start and end of the array.
        2. While `left` is less than `right`:
            a. Calculate the middle index `mid`.
            b. If the element at `mid` is greater than the element at `right`, it means the
               minimum element is in the right half of the array, so update `left` to `mid + 1`.
            c. Otherwise, the minimum element is in the left half (including `mid`), so
               update `right` to `mid`.
        3. The loop terminates when `left` equals `right`, and this index holds the minimum.
        4. Return the element at index `left`.
    """
    left, right = 0, len(nums) - 1  # Initialize left and right pointers
    while left < right:
        mid = (left + right) // 2  # Calculate the middle index
        if nums[mid] > nums[right]:
            left = mid + 1      # Minimum is in the right half
        else:
            right = mid          # Minimum is in the left half
    return nums[left]  # Return the minimum element



def find_min_recursive(nums: List[int], left: int, right: int) -> int:
    """
    Recursive binary search approach.

    This function finds the minimum element in a rotated sorted array using a recursive
    binary search algorithm.  It has the same time complexity of O(log n) as the iterative version.

    Args:
        nums (List[int]): A list of integers representing the rotated sorted array.
        left (int): The starting index of the subarray to search.
        right (int): The ending index of the subarray to search.

    Returns:
        int: The minimum element in the array.

    Algorithm:
        1. Base case: If `left` equals `right`, the subarray has only one element,
           which is the minimum, so return `nums[left]`.
        2. Calculate the middle index `mid`.
        3. If the element at `mid` is greater than the element at `right`, recursively
           search the right half of the array (from `mid + 1` to `right`).
        4. Otherwise, recursively search the left half of the array (from `left` to `mid`).
    """
    if left == right:
        return nums[left]  # Base case: single element
    mid = (left + right) // 2  # Calculate middle index
    if nums[mid] > nums[right]:
        return find_min_recursive(nums, mid + 1, right)  # Recursive call on right half
    else:
        return find_min_recursive(nums, left, mid)      # Recursive call on left half



def find_min_with_duplicates(nums: List[int]) -> int:
    """
    Binary search approach handling duplicates.

    This function handles rotated sorted arrays that may contain duplicate elements.
    The standard binary search approach may fail in the presence of duplicates, so this
    function includes a modification to handle them.  The time complexity can degrade to
    O(n) in the worst case (when all elements are the same), but is typically better.

    Args:
        nums (List[int]): A list of integers representing the rotated sorted array,
                           possibly with duplicates.

    Returns:
        int: The minimum element in the array.

    Algorithm:
        1. Initialize two pointers, `left` and `right`, to the start and end of the array.
        2. While `left` is less than `right`:
            a. Calculate the middle index `mid`.
            b. If the element at `mid` is greater than the element at `right`, the minimum
               is in the right half, so update `left` to `mid + 1`.
            c. If the element at `mid` is less than the element at `right`, the minimum
               is in the left half, so update `right` to `mid`.
            d. If the element at `mid` equals the element at `right`, it means we can't
               determine which half contains the minimum, so we reduce the search space
               by decrementing `right`.  This is the key modification to handle duplicates.
        3. Return the element at index `left`.
    """
    left, right = 0, len(nums) - 1
    while left < right:
        mid = (left + right) // 2
        if nums[mid] > nums[right]:
            left = mid + 1
        elif nums[mid] < nums[right]:
            right = mid
        else:
            right -= 1  # Handle duplicates: reduce search space
    return nums[left]



def find_min_linear(nums: List[int]) -> int:
    """
    Simple linear scan approach for robustness.

    This function finds the minimum element using a simple linear scan.  It iterates
    through the entire array and keeps track of the minimum element found so far.
    This approach is less efficient than binary search (O(n) time complexity), but it's
    very simple and robust, and it works correctly for any array, including those that
    are not sorted or rotated.

    Args:
        nums (List[int]): A list of integers.

    Returns:
        int: The minimum element in the array.

    Algorithm:
        1. Initialize a variable `min_val` to the first element of the array.
        2. Iterate through the rest of the array (from the second element to the last).
        3. For each element, compare it with `min_val`.  If the element is smaller than
           `min_val`, update `min_val` to the current element.
        4. After iterating through the entire array, `min_val` will hold the minimum element.
        5. Return `min_val`.
    """
    min_val = nums[0]  # Initialize with the first element
    for num in nums:
        if num < min_val:
            min_val = num  # Update min_val if a smaller element is found
    return min_val



def find_min_modified_pivot(nums: List[int]) -> int:
    """
    Modified pivot detection approach.

    This function is essentially the same as the iterative binary search approach
    (find_min_iterative).  It finds the minimum element by locating the "pivot" point
    in the rotated sorted array, which is the point where the array is rotated.
    The minimum element is located immediately after this pivot point.

    Args:
        nums (List[int]): A list of integers representing the rotated sorted array.

    Returns:
        int: The minimum element in the array.

    Algorithm:
        1. Initialize two pointers, `left` and `right`, to the start and end of the array.
        2. While `left` is less than `right`:
            a. Calculate the middle index `mid`.
            b. If the element at `mid` is greater than the element at `right`, it means the
               pivot point is in the right half, so update `left` to `mid + 1`.
            c. Otherwise, the pivot point is in the left half (including `mid`), so
               update `right` to `mid`.
        3. The loop terminates when `left` equals `right`, and this index holds the minimum.
        4. Return the element at index `left`.
    """
    left, right = 0, len(nums) - 1
    while left < right:
        mid = (left + right) // 2
        if nums[mid] > nums[right]:
            left = mid + 1
        else:
            right = mid
    return nums[left]



# Sample Test Cases
nums = [4, 5, 6, 7, 0, 1, 2]
print("Iterative Search:", find_min_iterative(nums))
print("Recursive Search:", find_min_recursive(nums, 0, len(nums) - 1))
print("Search with Duplicates:", find_min_with_duplicates([2, 2, 2, 0, 1]))
print("Linear Search:", find_min_linear(nums))
print("Modified Pivot Search:", find_min_modified_pivot(nums))

File: C++/5-BinarySearch/Problems/3-FindMinimum-in-RotatedSortedArray/FindMinimum-in-RotatedSortedArray.js
/**
 * Binary Search - Find Minimum in Rotated Sorted Array
 *
 * Problem:
 *
 * Given a sorted array that has been rotated n number of times, find the minimum element in it.
 *
 * Example:
 *
 * Input: arr[] = {5, 6, 1, 2, 3, 4}
 * Output: 1
 *
 * Input: arr[] = {1, 2, 3, 4}
 * Output: 1
 *
 * Input: arr[] = {3, 4, 5, 1, 2}
 * Output: 1
 *
 *
 * Constraints:
 *
 * -   The array is sorted but rotated.
 * -   The array has at least one element.
 * -   All elements are unique.
 *
 *
 * Follow-up:
 *
 * -   What is the time complexity of your solution?
 * -   Can you solve it in O(log n) time complexity?
 *
 */

/**
 * Approach 1: Linear Search
 *
 * -   Iterate through the array and find the minimum element.
 *
 * Time Complexity: O(n)
 * Auxiliary Space: O(1)
 */
function findMinLinearSearch(arr) {
  let min = arr[0];
  for (let i = 1; i < arr.length; i++) {
    if (arr[i] < min) {
      min = arr[i];
    }
  }
  return min;
}

/**
 * Approach 2: Using Math.min()
 *
 * -   Use the spread operator to expand the array and find the minimum element using Math.min().
 *
 * Time Complexity: O(n)
 * Auxiliary Space: O(1)
 */
function findMinMathMin(arr) {
  return Math.min(...arr);
}

/**
 * Approach 3: Binary Search - Iterative
 *
 * -   Use binary search to find the minimum element.
 * -   The minimum element is the only element whose previous element is greater than itself.
 * -   If the middle element is greater than the last element, then the minimum element is in the right half.
 * -   If the middle element is less than the last element, then the minimum element is in the left half.
 *
 * Time Complexity: O(log n)
 * Auxiliary Space: O(1)
 */
function findMinBinarySearchIterative(arr) {
  let low = 0;
  let high = arr.length - 1;

  while (low < high) {
    let mid = Math.floor((low + high) / 2);

    if (arr[mid] > arr[high]) {
      low = mid + 1;
    } else {
      high = mid;
    }
  }
  return arr[low];
}

/**
 * Approach 4: Binary Search - Recursive
 *
 * -   Use binary search recursively to find the minimum element.
 *
 * Time Complexity: O(log n)
 * Auxiliary Space: O(log n) (due to recursion stack)
 */
function findMinBinarySearchRecursive(arr, low, high) {
  if (low >= high) {
    return arr[low];
  }

  let mid = Math.floor((low + high) / 2);

  if (arr[mid] > arr[high]) {
    return findMinBinarySearchRecursive(arr, mid + 1, high);
  } else {
    return findMinBinarySearchRecursive(arr, low, mid);
  }
}

/**
 * Approach 5: Optimized Binary Search
 *
 * - A further optimized version of the iterative binary search.
 * - Handles cases where the array is not rotated at all.
 *
 * Time Complexity: O(log n)
 * Auxiliary Space: O(1)
 */
function findMinOptimized(arr) {
    let low = 0;
    let high = arr.length - 1;

    // Handle the case where the array is not rotated
    if (arr[low] <= arr[high]) {
        return arr[low];
    }

    while (low < high) {
        let mid = Math.floor((low + high) / 2);

        if (arr[mid] > arr[mid + 1]) {
            return arr[mid + 1];
        }
        if (arr[mid - 1] > arr[mid]) {
            return arr[mid];
        }

        if (arr[mid] > arr[low]) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    return -1; //Should not reach here, but added for safety.
}

// Example usage:
const arr1 = [5, 6, 1, 2, 3, 4];
const arr2 = [1, 2, 3, 4];
const arr3 = [3, 4, 5, 1, 2];
const arr4 = [10, 20, 30, 40, 50, 5, 6, 7];
const arr5 = [2, 1];
const arr6 = [2, 3, 4, 5, 6, 7, 8, 9, 10, 1];

console.log("Array 1:", arr1);
console.log("Linear Search:", findMinLinearSearch(arr1)); // Output: 1
console.log("Math.min():", findMinMathMin(arr1)); // Output: 1
console.log("Binary Search Iterative:", findMinBinarySearchIterative(arr1)); // Output: 1
console.log("Binary Search Recursive:", findMinBinarySearchRecursive(arr1, 0, arr1.length - 1)); // Output: 1
console.log("Optimized Binary Search:", findMinOptimized(arr1)); // Output: 1
console.log("\n");

console.log("Array 2:", arr2);
console.log("Linear Search:", findMinLinearSearch(arr2)); // Output: 1
console.log("Math.min():", findMinMathMin(arr2)); // Output: 1
console.log("Binary Search Iterative:", findMinBinarySearchIterative(arr2)); // Output: 1
console.log("Binary Search Recursive:", findMinBinarySearchRecursive(arr2, 0, arr2.length - 1)); // Output: 1
console.log("Optimized Binary Search:", findMinOptimized(arr2)); // Output: 1
console.log("\n");

console.log("Array 3:", arr3);
console.log("Linear Search:", findMinLinearSearch(arr3)); // Output: 1
console.log("Math.min():", findMinMathMin(arr3)); // Output: 1
console.log("Binary Search Iterative:", findMinBinarySearchIterative(arr3)); // Output: 1
console.log("Binary Search Recursive:", findMinBinarySearchRecursive(arr3, 0, arr3.length - 1)); // Output: 1
console.log("Optimized Binary Search:", findMinOptimized(arr3)); // Output: 1
console.log("\n");

console.log("Array 4:", arr4);
console.log("Linear Search:", findMinLinearSearch(arr4));
console.log("Math.min():", findMinMathMin(arr4));
console.log("Binary Search Iterative:", findMinBinarySearchIterative(arr4));
console.log("Binary Search Recursive:", findMinBinarySearchRecursive(arr4, 0, arr4.length - 1));
console.log("Optimized Binary Search:", findMinOptimized(arr4));
console.log("\n");

console.log("Array 5:", arr5);
console.log("Linear Search:", findMinLinearSearch(arr5));
console.log("Math.min():", findMinMathMin(arr5));
console.log("Binary Search Iterative:", findMinBinarySearchIterative(arr5));
console.log("Binary Search Recursive:", findMinBinarySearchRecursive(arr5, 0, arr5.length - 1));
console.log("Optimized Binary Search:", findMinOptimized(arr5));
console.log("\n");

console.log("Array 6:", arr6);
console.log("Linear Search:", findMinLinearSearch(arr6));
console.log("Math.min():", findMinMathMin(arr6));
console.log("Binary Search Iterative:", findMinBinarySearchIterative(arr6));
console.log("Binary Search Recursive:", findMinBinarySearchRecursive(arr6, 0, arr6.length - 1));
console.log("Optimized Binary Search:", findMinOptimized(arr6));
console.log("\n");
